airbnb_project/listings/tests/test_database_pipeline.py

```diff
@@ -0,0 +1,115 @@
+        item = {
+            'airbnb_listing_id': '12345',
+            'name': 'Test Listing',
+            'title': 'Beautiful Apartment',
+            'baths': '2',  # Converted to string
+            'beds': '3',  # Converted to string
+            'latitude': '49.2827',  # Converted to string
+            'longitude': '-123.1207',  # Converted to string
+            'person_capacity': '5',  # Converted to string
+            'registration_number': 'ABC123',
+            'room_type': 'Entire home/apt',
+            'location': 'Downtown Vancouver',
+            'bath_is_shared': False,  # Keep this as boolean
+            'baths_text': '2 baths'
+        }
+
+        # Process the item through the pipeline
+        self.pipeline.process_item(item, self.spider)
+
+        # Assert that a new listing was created in the database
+        self.assertEqual(Listing.objects.count(), 1)
+        listing = Listing.objects.get(airbnb_listing_id='12345')
+        self.assertEqual(listing.name, 'Test Listing')
+        self.assertEqual(listing.beds, '3')  # Update assertion to compare strings
+        self.assertEqual(listing.person_capacity, '5')  # Update assertion to compare strings
+
+    def test_process_item_skips_existing_listing(self):
+        # Create a listing in the database
+        Listing.objects.create(
+            airbnb_listing_id='12345',
+            name='Existing Listing',
+            title='Old Apartment',
+            baths='2',  # Converted to string
+            beds='3',  # Converted to string
+            latitude='49.2827',  # Converted to string
+            longitude='-123.1207',  # Converted to string
+            person_capacity='5',  # Converted to string
+            registration_number='DEF456',
+            room_type='Entire home/apt',
+            location='Downtown Vancouver',
+            is_bath_shared=False,  # Keep this as boolean
+            baths_text='2 baths'
+        )
+
+        # Simulate the same item scraped by the spider
+        item = {
+            'airbnb_listing_id': '12345',
+            'name': 'Test Listing',
+            'title': 'Beautiful Apartment',
+            'baths': '2',  # Converted to string
+            'beds': '3',  # Converted to string
+            'latitude': '49.2827',  # Converted to string
+            'longitude': '-123.1207',  # Converted to string
+            'person_capacity': '5',  # Converted to string
+            'registration_number': 'ABC123',
+            'room_type': 'Entire home/apt',
+            'location': 'Downtown Vancouver',
+            'bath_is_shared': False,  # Keep this as boolean
+            'baths_text': '2 baths'
+        }
+
+        # Process the item through the pipeline
+        self.pipeline.process_item(item, self.spider)
+
+        # Assert that the existing listing was not duplicated
+        self.assertEqual(Listing.objects.count(), 1)
+        listing = Listing.objects.get(airbnb_listing_id='12345')
```

Comment:
is this an actual database call? Shouldn't we be mocking this instead?

Reply:
This is using django's own test database functionality. Creates and destroys the db after the run.

Reply:
ah nice!! that's cool - we won't have to do mocking! If the tests are run in-parallel - we'd want to make sure that there is no race condition with any other test.